src/disk_io_thread.cpp

```diff
@@ -196,16 +197,23 @@ namespace libtorrent
 
 	void disk_io_thread::abort(bool wait)
 	{
-		m_abort = true;
-		if (num_threads() == 0)
+		// abuse the job mutex to make setting m_abort and checking the thread count atomic
+		// see also the comment in thread_fun
+		std::unique_lock<std::mutex> l(m_job_mutex);
+		if (m_abort.exchange(true)) return;
```

Comment:
given that m_job_mutex is held here, does m_abort still need to be an atomic? If all access to m_abort is done while holding the m_job_mutex, presumably it doesn't. If there are accesses to it where the mutex is not held, I take it those places would never interact with aborting jobs or exiting threads. It seems a bit suspicious though